addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java

```diff
@@ -0,0 +1,321 @@
+            }
+        }
+    }
+
+    /**
+     * Get new data from aqicn.org service
+     *
+     * @return {boolean}
+     */
+    private AirQualityJsonResponse updateAirQualityData() {
+        AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+        return getAirQualityData(StringUtils.trimToEmpty(config.getLocation()), config.getStationId());
+    }
+
+    /**
+     * Request new air quality data to the aqicn.org service
+     *
+     * @param location geo-coordinates from config
+     * @param stationId station ID from config
+     *
+     * @return the air quality data object mapping the JSON response or null in case of error
+     */
+    private AirQualityJsonResponse getAirQualityData(String location, Integer stationId) {
+        AirQualityJsonResponse result = null;
+        boolean resultOk = false;
+        String errorMsg = null;
+
+        try {
+
+            // Build a valid URL for the aqicn.org service
+            AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+
+            String geoStr = "geo:";
+            geoStr += location.replace(" ", "").replace(",", ";").replace("\"", "").replace("'", "").trim();
+
+            String urlStr = URL.replace("%APIKEY%", StringUtils.trimToEmpty(config.getApiKey()));
+
+            if (stationId == null) {
+                urlStr = urlStr.replace("%QUERY%", geoStr);
+            } else {
+                urlStr = urlStr.replace("%QUERY%", "@" + stationId);
+            }
+
+            logger.debug("URL = {}", urlStr);
+
+            // Run the HTTP request and get the JSON response from aqicn.org
+            URL url = new URL(urlStr);
+            URLConnection connection = url.openConnection();
+            String repsonse = IOUtils.toString(connection.getInputStream());
+            logger.debug("aqiResponse = {}", repsonse);
+
+            // Map the JSON response to an object
+            result = gson.fromJson(repsonse, AirQualityJsonResponse.class);
+            if (result == null) {
+                errorMsg = "no data returned";
+            } else if (result.getData() != null && result.getStatus() != "error") {
+                resultOk = true;
+            } else {
+                errorMsg = "missing data sub-object";
+            }
+
+            if (!resultOk) {
+                logger.warn("Error in aqicn.org (Air Quality) response: {}", errorMsg);
+            }
+
+        } catch (MalformedURLException e) {
+            errorMsg = e.getMessage();
+            logger.warn("Constructed url is not valid: {}", errorMsg);
+        } catch (JsonSyntaxException e) {
+            errorMsg = "Configuration is incorrect";
+            logger.warn("Error running aqicn.org (Air Quality) request: {}", errorMsg);
+        } catch (IOException | IllegalStateException e) {
+            errorMsg = e.getMessage();
+            logger.warn("Error running aqicn.org (Air Quality) request: {}", errorMsg);
```

Comment:
I would expect IOExceptions to happen if you do not have a network connection, right?
So according to the guidelines, you must not log a warning here, but set the Thing status accordingly only (what you do below).